### ISSUE TITLE
Batchnorm gives different results depending on whether cudnn is enabled

### ISSUE THREAD

**Issue by ssidorenko** [pytorch/pytorch#8283]
## Issue description

I've run into issues when training a model of mine. It works fine when disabling cuDNN.

I highly suspect the batchnorm layer to be the culprit. As seen in the code example below and the attached notebook, when cuDNN is enabled the results are inaccurate and non-deterministic.
## Code example
```
import torch
torch.backends.cudnn.enabled = False

bn = torch.nn.BatchNorm1d(1).cuda()
t = 100 * torch.ones([1,1,2]).cuda()
print(bn(t.cuda()))

# prints tensor([[[ 0.,  0.]]], device='cuda:0')

torch.backends.cudnn.enabled = True

bn = torch.nn.BatchNorm1d(1).cuda()
t = 100 * torch.ones([1,1,2]).cuda()
print(bn(t.cuda()))

# prints tensor(1.00000e-04 *       [[[ 8.5449,  8.5449]]], device='cuda:0')

```
## System Info
PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: 9.1.85

OS: Ubuntu 18.04 LTS
GCC version: (Ubuntu 5.5.0-12ubuntu1) 5.5.0 20171010
CMake version: version 3.9.4

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.1.85
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 390.48
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so
/usr/lib/x86_64-linux-gnu/libcudnn.so.7
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.1.2
/usr/lib/x86_64-linux-gnu/libcudnn_static.a

Versions of relevant libraries:
[pip3] numpy (1.14.2)
[pip3] numpydoc (0.7.0)
[pip3] pytorchviz (0.0.1)
[pip3] tensorboard-pytorch (0.7.1)
[pip3] torch (0.4.0)
[pip3] torchvision (0.2.0)
[conda] cuda91                    1.0                  h4c16780_0    pytorch
[conda] magma-cuda80              2.3.0                         1    pytorch
[conda] pytorch                   0.4.0           py36_cuda9.1.85_cudnn7.1.2_1  [cuda91]  pytorch
[conda] pytorchviz                0.0.1                     <pip>
[conda] tensorboard-pytorch       0.7.1                     <pip>
[conda] torchvision               0.2.1                    py36_1    pytorch

**Comment by vishwakftw** [comment]
If you want reproducibility, you could do this : `torch.backends.cudnn.deterministic = True`. By default, this is set to `False`.

**Comment by ngimel** [comment]
I was not able to reproduce non-deterministic results as long as bn weights are initialized to a fixed value. You are right that in your case the expected results is 0, as expected results is (x-mean)/var*weight[+bias], and, as long as mean is exactly equal to tensor values (as it should be), result is 0. However, due to some quirks of how cudnn computes output, a small (on the order of 1 ulp) error creeps in. With the default epsilon 1/var is approx 300, weight is on the order of 1, to get the result you are getting x-mean should be on the order of 1e-5. Given that x and mean are 100, it's enough for mean to have a 1e-7 relative error (which is approx 1 ulp) to produce the results you are seeing, which may happen due to fp arithmetic being inexact.
